python/ray/data/_internal/util.py

```diff
@@ -51,3 +58,97 @@ def _check_pyarrow_version():
             )
         else:
             _VERSION_VALIDATED = True
+
+
+def _autodetect_parallelism(
+    parallelism: int,
+    cur_pg: Optional["PlacementGroup"],
+    ctx: DatasetContext,
+    reader: Optional["Reader"] = None,
+    avail_cpus: Optional[int] = None,
+) -> (int, int):
+    """Returns parallelism to use and the min safe parallelism to avoid OOMs.
+
+    This detects parallelism based on the following heuristics (highest to lowest
```

Comment:
Maybe nit, but I feel like it's odd to say that this is highest to lowest precedence. The way I read this is, for example, that we would prefer to use num available CPUs even if it disrespects the max block size.

Reply:
Yeah, I changed this to just specify an order, hopefully it's better now.